### PR TITLE
feat(api): Allow recovering from errors that happen during the preparation part of an aspirate command

### PR DESCRIPTION
**Pull request by SyntaxColoring**
## Overview

Closes RQA-3567.

## Test Plan and Hands on Testing

* [ ] While the pipette is doing a plunger motion as part of preparation to aspirate, above the well, try inducing an overpressure error and recovering from it using the existing frontend.
  * I have not had any luck inducing an overpressure error here. It seems like the volume moved is too small.
* [x] Try inducing a stall error during the prepare-to-aspirate step, and recover from it by manually issuing a `commandType: "home"` `intent: "fixit"` command.

## Changelog

There are 4 substeps in an `aspirate` command:

1. Maybe, move above the well.
2. Maybe, move the plunger to recover from a prior `blowOut` and prepare for the `aspirate`.
3. Move to the target location (normally, into the liquid).
4. Move the plunger to aspirate the requested volume.

Prior work already lets us recover from overpressure and stall errors in steps 3 and 4. This does the same for steps 1 and 2.

## Review requests

Anyone have a trick to triggering an overpressure during the prepare-for-aspirate substep? Is it actually very difficult to trigger, and maybe this is barking up the wrong tree to solve RQA-3567?

## Risk assessment

Low.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/commands/aspirate.py [hunk 1]
```diff
@@ -102,14 +115,34 @@ async def execute(self, params: AspirateParams) -> _ExecuteReturn:
...
                 pipette_id=pipette_id,
                 labware_id=labware_id,
                 well_name=well_name,
                 well_location=WellLocation(origin=WellOrigin.TOP),
             )
+            state_update.append(move_result.state_update)
+            if isinstance(move_result, DefinedErrorData):
+                return DefinedErrorData(move_result.public, state_update=state_update)
+
+            # TODO: Figure out what to do for state_update_if_false_positive
```
Comment (sfoster1) [inline api/src/opentrons/protocol_engine/commands/aspirate.py]: imo passthrough and union with previous successes

File: api/src/opentrons/protocol_engine/state/update_types.py [hunk 2]
```diff
@@ -299,6 +299,19 @@ class StateUpdate:
...
+        This object is mutated in-place, taking values from `other`.
+        If an attribute in `other` is `NO_CHANGE`, the value in this object is kept.
+        """
+        fields = dataclasses.fields(other)
+        for field in fields:
+            other_value = other.__dict__[field.name]
+            if other_value != NO_CHANGE:
+                self.__dict__[field.name] = other_value
+        return self
+
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/update_types.py:314]: Is `append()` a good name for this? Is there a naming scheme that harmonizes better with `reduce()`?

File: api/src/opentrons/protocol_engine/state/update_types.py [hunk 3]
```diff
@@ -299,6 +299,19 @@ class StateUpdate:
...
     @classmethod
     def reduce(cls: typing.Type[Self], *args: Self) -> Self:
         """Fuse multiple state updates into a single one.
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/state/update_types.py:317]: Note to self: Can probably implement `reduce()` in terms of `append()`.

File: api/src/opentrons/protocol_engine/commands/aspirate.py (hunk 1 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/commands/aspirate.py]: I think I've settled on ignoring `state_update_if_false_positive` until/unless this command actually uses it for something, in which case the right thing to do will hopefully be more obvious.

The reason I'm not setting up a pattern of "passthrough and union with previous successes" (yet) is just uncertainty on my part. Like, it kind of smells to me like an `Implementation.execute()` concern that our inner helper functions aren't in a good position to anticipate. And I wouldn't want to go out of our way to set up a pattern that does nothing now and starts doing the wrong thing later on.

File: api/src/opentrons/protocol_engine/state/update_types.py (hunk 2 above)
Comment (sfoster1) [reply api/src/opentrons/protocol_engine/state/update_types.py:314]: you could make _this_ `reduce()`, continue to take N arguments, and have self as a silent first argument. that is probably the most natural.

**Review by sfoster1 — APPROVED** [review]
Looks good, minor name suggestion, but lots of extra work probably

**Review by TamarZanzouri — APPROVED** [review]
Nice!

**Pull request merged**